flb_encoding: charset encoding for input plugins

bluebike opened the pull request:
<!-- Provide summary of changes -->

Adds library flb_encoding for doing charset encoding   CHARSET => UTF8. 

* Uses lib/tutf8e-library. 
* Only 8-bit charsets are supported.
* At first *in_tail* plugin is supported, later in_syslog and others. 

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

bluebike committed:
encoding: add flb_encoding

Add flb_encoding functions for charset encodings to utf8.

* Uses lib/tutf8e-library.
* Only 8-bit source charsets are supported.
* Support for replacement string (in case of bad chars)
* This commit doesn't add support to any input plugin.

Signed-off-by: Jukka Pihl <jukka.pihl@iki.fi>

bluebike committed:
in_tail: add flb_encoding support to in_tail-plugin

* Adds new option:  "encoding" to in_tail.
* If encoding fails. message is skipped.

Signed-off-by: Jukka Pihl <jukka.pihl@iki.fi>

bluebike commented:
I build this over #2326 (those commits seems to be included in PR), so I would like that to be merged before this. 
TODO:  example run, configuration,  valgrind.
@nigels-com  @edsiper   would this be usable?

edsiper commented:
#2326 was just merged, @bluebike @nigels-com can you remove the tutf8e pieces from this PR ?

in addition, ideally we want separate commits for the core interface "encoding" and for the plugins being improved.

bluebike commented:
Rebased on master (with #2326 merged).
In two commits: 
1) flb_encoding
2) in_tail changes

TODO(?)
- [ ]  add in_syslog support
- [ ]  test example

bluebike committed:
in_syslog: support for input encoding to utf8

* encoding uses flb_encoding
* refactoring syslog_prot_process and syslog_prot_process_udp to
use syslog_prot_process_msg per message.

Signed-off-by: Jukka Pihl <jukka.pihl@iki.fi>

bluebike commented:
Added encoding support to  **in_syslog** . 
Compiling with FLB_UTF8_ENCODER=No  worked also... didn't see any related warnings in my env (macOS 10..13).

edsiper commented on src/flb_encoding.c:
```diff
@@ -0,0 +1,138 @@
+        opt += 2;
+        if (*opt == '/') {
+            invalid = opt + 1;
+        } else if (!strcmp(opt,"I") || !strcmp(opt,"IGNORE")) {
```
```else if ``` must be in a new line: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

bluebike replied:
ok... let's fix those...

edsiper commented on src/flb_encoding.c:
```diff
@@ -0,0 +1,138 @@
+        flb_error("[flb_encoding] unknown encoding: %s", encoding);
+        return NULL;
+    }
+    ec =   flb_calloc(sizeof(struct flb_encoding),1);
```
validate that ```ec``` is not null, if so trigger an flb_errno() and abort

bluebike replied:
yes....

bluebike replied:
added also check for **flb_strdup**(invalid)

edsiper commented on src/flb_encoding.c:
```diff
@@ -0,0 +1,138 @@
+ *  windows-1251 windows-1252, ..
+ * 
+ *  <charset>                     - fail if bad chars
+ *  <charset>//IGNORE             - ignore bad chars
```
is the ```//``` just a separator or a common naming across encoding configs ?

bluebike replied:
iconv(3)   library uses same convention to add  parameters to  charset.
implementations I know understand only   //IGNORE  and //TRANSLIT .

nigels-com replied:
I'm a bit sceptical of following the iconv API for this.
I think a nullable _invalid_ UTF8 string is sufficient for configuration purposes.
I don't think fluent-bit should be concerned with //TRANSLIT mode in particular:

```
ICONV_OPEN(3)                                               Linux Programmer's Manual                                               
...
NAME
       iconv_open - allocate descriptor for character set conversion

...
       //TRANSLIT
              When the string "//TRANSLIT" is appended to tocode, transliteration is activated.  This means that when a character cannot be  rep‐
              resented in the target character set, it can be approximated through one or several similarly looking characters.

       //IGNORE
              When  the  string  "//IGNORE"  is  appended  to  tocode,  characters that cannot be represented in the target character set will be
              silently discarded.```

bluebike replied:
//TRANSLIT has no meaning in our case. 
//IGNORE is usefull...

bluebike replied:
So ... we are going to get some kind agreement of configuration?

I like idea of one configuration value with //(parameter), because that is flexible and  "kind of"  compatible with iconv.
Also it's useful to have multiple ways to handle bad input (ignore,fail,replace-with-something).

nigels-com replied:
I don't think the iconv formulation is particularly modern, concise or appropriate for `fluent-bit` (in my opinion).

bluebike replied:
So...  what this "modern" could mean?

bluebike replied:
So how this should be configured???????

nigels-com replied:
I propose that the plugin supports an optional UTF8 _invalid_ string, along the lines of tutf8e.
//IGNORE would simply be an empty string, default might be ? or similar.

bluebike replied:
Well...  that is this code doing... just putting everything to one parameter (like but not totally same as in iconv(3)). 
But should this be an another parameter????

Then probably better that replacement is should be (json) escaped string.
`

Encoding  iso-8859-2               => default... FAIL,INGORE ,  "?", "\uFFFD"  ????
Encoding  iso-8859-2  ?            => 
Encoding  iso-8859-3  <bad\20char> => "<bad char
Encoding  iso-8859-3  \uFFFD       => Use unicode replacement character

Maybe  

Encoding  iso-8859-2  \I            =>  IGNORE
Encoding  iso-8859-2  \F            =>  FAIL
Encoding  iso-8859-2  \R            =>  Replacement char...
`
Or we put optional second parameter to quotes "?" ...
hum... getting complicated.

bluebike replied:
Are getting anywhere with configuration????  @nigels-com

nigels-com replied:
I'll take a fresh look.  I know you came to this from an iconv perspective, I was hoping to leave that behind rather than carry that into fluent-bit.  It's been a while since I was actively engaged on this one.

bluebike replied:
hum.. maybe have to change way of configuration...

bluebike replied:
Changed configuration  use have parameters: 

* **encoding**  CHARSET
* **encoding_replacement**  STRING  

default is to use unicode replacement char (0xfffd)

edsiper reviewed (changes requested):
thanks. I wrote some comments.

About the second commit, note that it must be prefixed with ```in_tail: .....```

bluebike commented:
Did  requested changes... 
* checking memory allocation
* formatting checked. 
(... I'll add test examples soon)

bluebike commented:
**in_tai**l:   simple test run in shell.   input contains  ä,Ö and €  characters encoded in windows-1252 (cp1252). 

```

$  echo $'Test data'    > huuhaa.txt
$  echo $'This contains a+dots: \xe4 O+dots: \xd6. trailing data' >> huuhaa.txt
$  echo $'This contains euro character: \x80' >> huuhaa.txt



$   bin/fluent-bit -v -i tail -p path=huuhaa.txt -p 'encoding=windows-1252'  -o stdout

Fluent Bit v1.6.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/08/11 22:06:42] [ info] Configuration:
[2020/08/11 22:06:42] [ info]  flush time     | 5.000000 seconds
[2020/08/11 22:06:42] [ info]  grace          | 5 seconds
[2020/08/11 22:06:42] [ info]  daemon         | 0
[2020/08/11 22:06:42] [ info] ___________
[2020/08/11 22:06:42] [ info]  inputs:
[2020/08/11 22:06:42] [ info]      tail
[2020/08/11 22:06:42] [ info] ___________
[2020/08/11 22:06:42] [ info]  filters:
[2020/08/11 22:06:42] [ info] ___________
[2020/08/11 22:06:42] [ info]  outputs:
[2020/08/11 22:06:42] [ info]      stdout.0
[2020/08/11 22:06:42] [ info] ___________
[2020/08/11 22:06:42] [ info]  collectors:
[2020/08/11 22:06:42] [ info] [engine] started (pid=47950)
[2020/08/11 22:06:42] [debug] [engine] coroutine stack size: 12288 bytes (12.0K)
[2020/08/11 22:06:42] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/08/11 22:06:42] [ info] [storage] version=1.0.5, initializing...
[2020/08/11 22:06:42] [ info] [storage] in-memory
[2020/08/11 22:06:42] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/08/11 22:06:42] [debug] [input:tail:tail.0] scanning path huuhaa.txt
[2020/08/11 22:06:42] [debug] [input:tail:tail.0] inode=13109344 appended as huuhaa.txt
[2020/08/11 22:06:42] [debug] [input:tail:tail.0] scan_glob add(): huuhaa.txt, inode 13109344
[2020/08/11 22:06:42] [debug] [input:tail:tail.0] 1 new files found on path 'huuhaa.txt'
[2020/08/11 22:06:42] [debug] [router] default match rule tail.0:stdout.0
[2020/08/11 22:06:42] [ info] [sp] stream processor started
[2020/08/11 22:06:42] [debug] [input:tail:tail.0] inode=13109344 file=huuhaa.txt promote to TAIL_EVENT
[2020/08/11 22:06:47] [debug] [task] created task=0x7f9b36d00a30 id=0 OK
[0] tail.0: [1597172802.983562000, {"log"=>"Test data"}]
[1] tail.0: [1597172802.983579000, {"log"=>"This contains a+dots: ä O+dots: Ö. trailing data"}]
[2] tail.0: [1597172802.983580000, {"log"=>"This contains euro character: €"}]
^C[engine] caught signal (SIGINT)
[2020/08/11 22:06:52] [ info] [input] pausing tail.0
[2020/08/11 22:06:52] [debug] [input:tail:tail.0] inode=13109344 removing file name huuhaa.txt

```

bluebike commented:
in_syslog:  test run using UDP syslog messages.

```
# start fluent-bit first.. then send these in different terminal (+ bash shell)

$ echo $'<135>Aug 11 20:27:22 myhost test: nothing'   | nc -w 1 -u 127.0.0.1  7700
$ echo $'<135>Aug 11 20:27:22 myhost test: euro: \x80'   | nc -w 1 -u 127.0.0.1  7700
$ echo $'<135>Aug 11 20:27:22 myhost test:  tama: t\xe4m\xe4'   | nc -w 1 -u 127.0.0.1  7700



$ bin/fluent-bit -v -R ../conf/parsers.conf -i syslog -p mode=udp  -p port=7700 -p Parser=syslog-rfc3164-local  -p encoding=windows-1252 -o stdout

Fluent Bit v1.6.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/08/11 22:13:51] [ info] Configuration:
[2020/08/11 22:13:51] [ info]  flush time     | 5.000000 seconds
[2020/08/11 22:13:51] [ info]  grace          | 5 seconds
[2020/08/11 22:13:51] [ info]  daemon         | 0
[2020/08/11 22:13:51] [ info] ___________
[2020/08/11 22:13:51] [ info]  inputs:
[2020/08/11 22:13:51] [ info]      syslog
[2020/08/11 22:13:51] [ info] ___________
[2020/08/11 22:13:51] [ info]  filters:
[2020/08/11 22:13:51] [ info] ___________
[2020/08/11 22:13:51] [ info]  outputs:
[2020/08/11 22:13:51] [ info]      stdout.0
[2020/08/11 22:13:51] [ info] ___________
[2020/08/11 22:13:51] [ info]  collectors:
[2020/08/11 22:13:51] [ info] [engine] started (pid=48082)
[2020/08/11 22:13:51] [debug] [engine] coroutine stack size: 12288 bytes (12.0K)
[2020/08/11 22:13:51] [debug] [storage] [cio stream] new stream registered: syslog.0
[2020/08/11 22:13:51] [ info] [storage] version=1.0.5, initializing...
[2020/08/11 22:13:51] [ info] [storage] in-memory
[2020/08/11 22:13:51] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/08/11 22:13:51] [ info] [in_syslog] UDP buffer size set to 32768 bytes
[2020/08/11 22:13:51] [ info] [in_syslog] UDP server binding 0.0.0.0:7700
[2020/08/11 22:13:51] [debug] [router] default match rule syslog.0:stdout.0
[2020/08/11 22:13:51] [ info] [sp] stream processor started
[0] syslog.0: [1597177642.000000000, {"pri"=>"135", "time"=>"Aug 11 20:27:22", "ident"=>"myhost", "message"=>"nothing"}
[0] syslog.0: [1597177642.000000000, {"pri"=>"135", "time"=>"Aug 11 20:27:22", "ident"=>"myhost", "message"=>"euro: €"}]
[0] syslog.0: [1597177642.000000000, {"pri"=>"135", "time"=>"Aug 11 20:27:22", "ident"=>"myhost", "message"=>"tama: tämä"}]
```

nigels-com commented on src/flb_encoding.c:
```diff
@@ -0,0 +1,159 @@
+            return NULL;
+        }
+    }
+    else {
```
Recommend a documentation PR to go along with this over at: 
https://github.com/fluent/fluent-bit-docs

bluebike replied:
Documentation  for change:

https://github.com/fluent/fluent-bit-docs/pull/410

nigels-com commented on src/flb_encoding.c:
```diff
@@ -0,0 +1,159 @@
+void flb_encoding_close(struct flb_encoding *ec) {
+    if (ec) {
+        if (ec->invalid) {
+            flb_free((char*)ec->invalid);
```
_invalid_ isn't always heap allocated?  Looks like a potential crash here.

bluebike replied:
if invalid is not NULL then it's always flb_strdup()ed

bluebike replied:
there was possiblity that  _struct flb_encoding_ was allocated but could not allocate _invalid_. 
Fixed that.

bluebike commented:
Added documentation PR https://github.com/fluent/fluent-bit-docs/pull/410
and fixed possible memory leak if allocation fails in opening in `flb_encoding_open`

nigels-com commented on plugins/in_tail/tail.c:
```diff
@@ -613,6 +613,14 @@ static struct flb_config_map config_map[] = {
+    {
+     FLB_CONFIG_MAP_STR, "encoding", NULL,
+     0, FLB_FALSE, 0,
+     "specify the charset encoder to decode message",
```
Perhaps to clarify the wording for the encoding parameter:
```specify the input encoding for converting to UTF-8```

bluebike replied:
Changed text tail.c test to `specify the input encoding for converting to UTF-8`

bluebike commented:
> thanks. I wrote some comments.
> 
> About the second commit, note that it must be prefixed with `in_tail: .....`

Changed commit message.

kingjan1999 commented:
Any news on this? Would love to see this merged!

edsiper commented:
assigned to @nokute78 for review

hpernu commented:
Could we get this included in mainline ASAP? I assume I am not the only one with non-UTF8 log entries . 

As a user, I do not even really care how this is implemented but as for configuration, input seems the most convenient i.e. tail-plugin.

cosmo0920 commented:
This PR is superseded by https://github.com/fluent/fluent-bit/pull/10464 and https://github.com/fluent/fluent-bit/pull/10542.
